src/attr.rs

```diff
@@ -24,12 +27,90 @@ pub fn valid<I: Iterator<Item = char>>(chars: I) -> (usize, bool) {
     (p.pos, has_attr)
 }
 
+/// Stores an attribute value that supports backslash escapes of ASCII punctuation upon displaying,
+/// without allocating.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AttributeValue<'s> {
+    raw: CowStr<'s>,
+}
+
+impl<'s> AttributeValue<'s> {
+    /// Processes the attribute value escapes and returns an iterator of the parts of the value
+    /// that should be displayed.
+    pub fn parts(&'s self) -> AttributeValueParts<'s> {
+        let mut split = self.raw.split('\\');
+        let next = split.next();
+        AttributeValueParts {
+            next,
+            split,
+            escape: true,
+        }
+    }
+}
+
+impl<'s> From<&'s str> for AttributeValue<'s> {
+    fn from(value: &'s str) -> Self {
+        Self { raw: value.into() }
+    }
+}
+
+impl<'s> From<CowStr<'s>> for AttributeValue<'s> {
+    fn from(value: CowStr<'s>) -> Self {
+        Self { raw: value }
+    }
+}
+
+impl<'s> From<String> for AttributeValue<'s> {
+    fn from(value: String) -> Self {
+        Self { raw: value.into() }
+    }
+}
+
+impl<'s> fmt::Display for AttributeValue<'s> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.parts().try_for_each(|part| f.write_str(part))
+    }
+}
+
+/// An iterator over the parts of an [`AttributeValue`] that should be displayed.
+pub struct AttributeValueParts<'s> {
+    split: Split<'s, char>,
+    escape: bool,
+    next: Option<&'s str>,
```

Comment:
Two thoughts about this iterator:

- the size seems quite large, it is currently 96 bytes, not sure how the Split
  looks on the inside but it seems surprisingly large.
- currently it seems to return e.g. `"a" + "\" + "b"` for `"a\\b"`, but ideally
  it should return `"a" + "\b"`, right?

I'm thinking it should be possible with a single &'str (16 bytes). E.g for the
string `"a\.b\c\\."` we get three strs emitted:

- new: `ahead <= "a\.b\c\\."`,
- next: first backslash has punctuation after
    => `ahead <= ".b\c"`, return `Some("a")`
- next: second backslash has punctuation after,
    => `ahead <= "\."`, return `Some(".b\c")`
- next: backslash at pos 0, so we assume it has already been escaped
    => `ahead <= ""`, return `Some("\.")`
- next: ahead is empty
    => return None

In the case where it starts with a backslash:

- `"\abc"` => new: `ahead <= "\abc"`, so backslash gets emitted,
- `"\.abc"` => new: `ahead <= ".abc"`, so the backslash does not get emitted,

It's not necessary to try this out right now, though.

Reply:
I have to say; you are a wizard! Not only did that method reduce the memory footprint of the iterator, I also found a way to split the string into as few as possible splits while implementing it. All tests still pass, and this is definitely the way to go!